service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+package frontend
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/status"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.temporal.io/api/proxy"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/rpc/encryption"
+	"go.temporal.io/server/common/util"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+)
+
+// HTTPAPIServer is an HTTP API server that forwards requests to gRPC via the
+// gRPC interceptors.
+type HTTPAPIServer struct {
+	server                 http.Server
+	listener               net.Listener
+	logger                 log.Logger
+	serveMux               *runtime.ServeMux
+	shutdownDrainTime      time.Duration
+	stopped                chan struct{}
+	matchAdditionalHeaders map[string]bool
+}
+
+var defaultForwardedHeaders = []string{
+	"Authorization-Extras",
+	"X-Forwarded-For",
+	http.CanonicalHeaderKey(headers.ClientNameHeaderName),
+	http.CanonicalHeaderKey(headers.ClientVersionHeaderName),
+}
+
+type httpRemoteAddrContextKey struct{}
+
+// NewHTTPAPIServer creates an [HTTPAPIServer].
+func NewHTTPAPIServer(
+	serviceConfig *Config,
+	rpcConfig config.RPC,
+	grpcListener net.Listener,
+	tlsConfigProvider encryption.TLSConfigProvider,
+	handler Handler,
+	interceptors []grpc.UnaryServerInterceptor,
+	logger log.Logger,
+) (*HTTPAPIServer, error) {
+	// Create a TCP listener the same as the frontend one but with different port
+	tcpAddrRef, _ := grpcListener.Addr().(*net.TCPAddr)
+	if tcpAddrRef == nil {
+		return nil, fmt.Errorf("must use TCP for gRPC listener to support HTTP API")
+	}
+	tcpAddr := *tcpAddrRef
+	tcpAddr.Port = rpcConfig.HTTPPort
+	if tcpAddr.Port == 0 {
+		tcpAddr.Port = tcpAddrRef.Port + 10
```

Comment:
Adding 10 here to the RPC port might work for our static configs, but it could conflict with others--it's also just too much implicit behavior IMO.

Reply:
Are you suggesting I open on `7243` by default regardless of your gRPC port? I can do that, though would like others' opinions.

Reply:
The +10 feels too magical and surprising to me also. I'd rather just require an explicit port here. All port selection should be done in static config (or by something wrapping the server and generating a static config in memory)

Feel free to add an explicit port to all the development configs and dev servers

Reply:
Do we require an explicit port for gRPC today? What does the absence of a RPC port in config mean today?

Reply:
Great! That's information I didn't know. Yes, I will make this required and set the default in static configs. I assume this is only `docker/config_template.yaml`? Can you confirm for me what happens if frontend gRPC port config is left off? (I can try to dig as well, but I assume it picks a random port which makes it not very usable)

Reply:
This has been fixed to require an HTTP port since we require it for gRPC. The absence of an HTTP port (or 0) means disabled. Docker config templates have been updated to set to 7243 by default.